### PR TITLE
turn off specialize float

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137782

This is the next step in support dynamic float arguments in PT2: https://docs.google.com/document/d/1HswUSp9H6mg8Vg27mhRk8YzC9q_uf63b6wz-gwx65BQ/edit?pli=1#heading=h.xvyiqp8tuje6. To make this more incremental and tractable, we've decided to opt the export path our of this first phase of the rollout.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

### CODE REVIEW

File: torch/fx/passes/_tensorify_python_scalars.py [hunk 1]
```diff
@@ -248,6 +250,44 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
...
+    for i, node in enumerate(nodes[:-1]):
+        with graph.inserting_before(
+            nodes[i + 1] if node not in placeholders else first_non_placeholder
+        ):
+            args = []
+            transform = False
+            for a in node.args:
+                if (
+                    isinstance(a, fx.Node)
+                    and "val" in a.meta
```
Comment (bobrenjc93) [inline torch/fx/passes/_tensorify_python_scalars.py:265]: I had to be more defensive here b/c tensor constants don't have "val" set in meta

eg.

```
b = torch.tensor(1.0, dtype=torch.double, device=self.device)
```

`python test/inductor/test_torchinductor.py CpuTests.test_clamp_type_promotion_cpu`

doesn't have "val" set in meta and results in a KeyError

File: test/dynamo/test_repros.py [hunk 2]
```diff
@@ -5708,6 +5708,7 @@ def fn(x, y):
 
     @torch._dynamo.config.patch(guard_nn_modules=False)
     @torch._dynamo.config.patch(inline_inbuilt_nn_modules=False)
+    @torch._dynamo.config.patch(specialize_float=True)
```
Comment (ezyang) [inline test/dynamo/test_repros.py:5711]: comment why

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -185,6 +185,7 @@ def _check_tensorbox(nodes: Optional[_NodeOrNodes]) -> None:
                     sympy.logic.boolalg.Boolean,
                     Expr,
                     int,
+                    float,
```
Comment (ezyang) [inline torch/_inductor/ir.py:188]: If we always spec when we fail to eliminate SymFloat, we shouldn't actually hit this case of the assert, right?

File: torch/fx/interpreter.py [hunk 4]
```diff
@@ -271,6 +272,17 @@ def call_function(self, target : 'Target', args : Tuple[Argument, ...], kwargs :
...
+        if not torch._dynamo.config.specialize_float:
+            # Force the specialization of backed symfloats
+            args = tuple(
+                float(a) if isinstance(a, torch.SymFloat) and isinstance(a.node.expr, Symbol) and a.node.hint is not None else a
+                for a in args
+            )
+            kwargs = {
+                k: (float(v) if isinstance(v, torch.SymFloat) and isinstance(v.node.expr, Symbol) and v.node.hint is not None else v)
+                for k, v in kwargs.items()
+            }
```
Comment (ezyang) [inline torch/fx/interpreter.py:284]: I agree with your analysis in the doc that this is the wrong place to put it. But for mode="eager" it's kind of strange things are failing, because the eager interpreter should be able to handle floats directly, so we don't have to avoid specializing in this case. Hard to have a better idea without more info on the tests that were failing

Comment (bobrenjc93) [reply torch/fx/interpreter.py:284]: So I put up https://github.com/pytorch/pytorch/pull/138433/files which is basically this diff with the interpreter stuff ripped out.

Here's an example test that fails that diff but passes this one

```
python test/inductor/test_torchinductor.py CpuTests.test_baddbmm_cpu
```

If only there was a way to do set operations on tests across PRs! Then I could quickly get you a list of tests that pass only when this wonky interpreter change is in. I guess it's time to start that side quest ;)

Comment (bobrenjc93) [reply torch/fx/interpreter.py:284]: I put up https://github.com/pytorch/pytorch/pull/138394 to start logging test failure repros. I didn't know how to test it locally so I just pushed a PR and will check to see if it works in a few hours. Fingers crossed it actually logs correctly!

Comment (ezyang) [reply torch/fx/interpreter.py:284]: test_cauchy_cuda_float64 looks fun. It sounds like when we started capturing the float we ended up with incorrect semantics in the recorded graph. Shouldn't be difficult to try to solve for directly?

File: torch/_dynamo/guards.py [hunk 5]
```diff
@@ -338,6 +339,7 @@ def _get_closure_vars():
             "___as_tensor": torch._as_tensor_fullprec,
             "torch": torch,
             "inspect": inspect,
+            "OpaqueUnaryFn_sqrt": OpaqueUnaryFn_sqrt,
```
Comment (ezyang) [inline torch/_dynamo/guards.py:342]: This shouldn't be here, and binding it directly to OpaqueUnaryFn_sqrt is also not right. It should go in SYMPY_INTERP in torch.fx.experimental.symbolic_shapes and the rhs should be math.sqrt

File: torch/_dynamo/symbolic_convert.py [hunk 6]
```diff
@@ -550,6 +550,11 @@ def inner(self: "InstructionTranslatorBase", inst: Instruction):
                         if push:
                             self.push(value)
                         self.jump(inst)
+                elif isinstance(result, SymNodeVariable):
+                    if result.evaluate_expr():
+                        if push:
+                            self.push(value)
+                        self.jump(inst)
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:557]: Can land by itself if you want

File: torch/_dynamo/variables/tensor.py [hunk 7]
```diff
@@ -1148,9 +1148,16 @@ def as_proxy(self):
...
         if self._tensor_var is None:
-            self._tensor_var = VariableTracker.build(
-                tx, torch.scalar_tensor
-            ).call_function(tx, [self], {})
+            if isinstance(self.sym_num, torch.SymFloat):
+                self._tensor_var = VariableTracker.build(
+                    tx, torch.scalar_tensor
+                ).call_function(
+                    tx, [self], {"dtype": VariableTracker.build(tx, torch.float64)}
+                )
```
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:1156]: hmmmmm. Maybe we should just unconditionally set dtype for everything here

Comment (ezyang) [reply torch/_dynamo/variables/tensor.py:1156]: Nice catch!

File: torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py [hunk 8]
```diff
@@ -130,9 +130,24 @@ def aot_dispatch_base_graph(
             mod_when_exporting_non_strict, assigned_buffers
         )
 
-    saved_updated_flat_args_subclasses_desugared = pytree.tree_map_only(
-        torch.Tensor, lambda t: t.detach(), updated_flat_args_subclasses_desugared
```
Comment (ezyang) [inline torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:134]: A better implementation would be to have detach propagate memos

File: torch/_subclasses/fake_tensor.py [hunk 9]
```diff
@@ -560,9 +560,13 @@ def _memo_epoch(self, obj: FakeTensor) -> str:
 
     def __get__(
         self, obj: FakeTensor, objtype: Optional[Type[FakeTensor]] = None
-    ) -> Optional[torch.SymInt]:
+    ) -> Optional[Union[torch.SymInt, torch.SymFloat]]:
         if (r := getattr(obj, self._memo(obj))) is None:
             return None
+
+        if isinstance(r, torch.SymFloat):
+            return r
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:568]: Memo is unconditionally ignored for SymFloat memos

Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:568]: A tighter condition probably involves testing if something has any unbacked things in it, because if there aren't any (and with floats you wouldn't expect there to be e.g., zf0), then that's when it's OK to preserve the memo (because it won't renumber)

File: torch/fx/experimental/symbolic_shapes.py [hunk 10]
```diff
@@ -442,6 +442,11 @@ def rebind_unbacked(
         assert shape_env is not None
         for raw_u0, path in bindings.items():
             u1 = pytree.key_get(result, path)
+
+            # We only care about rebinding unbacked SymInts
+            if not isinstance(u1, torch.SymInt):
+                continue
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:448]: You care about rebinding ALL unbacked things, but many of your SymFloats are not unbacked. The better test is just to check if there are any unbacked symbols in u1.

File: torch/fx/experimental/symbolic_shapes.py [hunk 11]
```diff
@@ -6082,6 +6087,12 @@ def _evaluate_expr(
 
         # TODO: split conjunctions and evaluate them separately
 
+        if isinstance(
+            orig_expr,
+            (sympy.logic.boolalg.BooleanTrue, sympy.logic.boolalg.BooleanFalse),
+        ):
+            return orig_expr
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:6094]: This is fine but it's weird that we never ran into this before. In particular, I would have expected there to be some preexisting logic to circuit out when orig_expr is just a constant.

File: torch/fx/passes/split_module.py [hunk 12]
```diff
@@ -281,7 +281,7 @@ def instantiate_node_partition_mapping(node):
         # See https://github.com/pytorch/pytorch/issues/130534
         if (
             (val := node.meta.get("example_value")) is not None and
-            isinstance(val, torch.SymInt) and
+            isinstance(val, (torch.SymInt, torch.SymFloat)) and
```
Comment (ezyang) [inline torch/fx/passes/split_module.py:284]: good
